blog-post/blog-post.md

```diff
@@ -0,0 +1,634 @@
+Rust traits: they provide signatures for the functions which will be implemented, for
+distinct types, in instance definitions. In this sense, instance declarations are similar
+to `impl` in Rust.
+
+A type class definition declares the class name, its arguments and member functions type
+signatures. As example, let's consider the following definition of a class for the ABI
+encoding of types:
+
+```
+forall self . class self:ABIEncode {
+    function encodeInto(x:self, basePtr:word, offset:word, tail:word) -> word;
+}
+```
+
+Function `encodeInto` encodes a value of type `self` into a memory region starting at
+`basePtr`. Argument `offset` denotes the position of the first empty byte of the head
+and `tail` which gives the position of the first empty byte of the tail. The
+implementation of encoding for a specific type is done in an instance declaration.
+As an example, the following instance definition implements the ABI encoding for
+`uint256` type, which is represented as an algebraic type that holds a `word` value.
+
+```
+data uint256 = uint256(word);
+instance uint256:ABIEncode {
+    function encodeInto(x:uint256, basePtr:word, offset:word, tail:word) -> word {
+        let repx : word = Typedef.rep(x);
+        assembly { mstore(add(basePtr, offset), repx) }
+        return tail;
+    }
+}
+```
+Type classes and instances can be used to avoid repetitive definitions.
+As an example, consider the [console](https://github.com/foundry-rs/forge-std/blob/master/src/console.sol) library
+present in the foundry framework. Looking the code, we can see
+that there are several one argument functions which only call the ABI
+encoding for the value being logged and then perform a static call to
+the console address. All these one argument functions definitions can
+be reduced to following single function, which uses the standard library
+machinery for ABI encoding of values.
+
+```
+forall ty . ty : ABIAttribs, ty : ABIEncode => function log(val : ty) {
+    let CONSOLE_ADDRESS : word = 0x000000000000000000636F6e736F6c652e6c6f67;
+    let payload = abi_encode(val);
+
+    // extract the underlying word representation of the payload
+    let ptr = Typedef.rep(payload);
+
+    assembly {
+        pop(
+            staticcall(
+                gas(),
+                CONSOLE_ADDRESS,
+                add(ptr, 32),
+                mload(ptr),
+                0,
+                0
+            )
+        )
+    }
+}
+```
+
+Another short possible implementation which relies on EVM `log1` instruction,
+instead of Foundry console address, would be:
+
+```
+forall t . t : Typedef(word) => function log1(v : t, topic : word) -> () {
+    let w : word = Typedef.rep(v);
+    assembly {
+        mstore(0,w)
+        log1(0,32,topic)
+    }
+}
+```
+
+This version uses standard library class `Typedef`, which provides functions for
+the conversion between `word` and types which are instance of this class.
```

Comment:
I think it would be helpful to point out that `t : Typedef(word)` puts a constraint on which types can be used (so that only ones that actually have `word` representation will not produce an error). Similarly in the encoding example I'd mention that it's limited to ABI-encodable types.

Reply:
ABI encoding is removed from this section and now lives in the section on SAIL. I kept console.log and added Typedef to this section since it was needed for console.log.